### ISSUE TITLE
Rule JJ01 doesn't work for directories in "parallel processing" mode

### ISSUE THREAD

**Issue by shr1k** [sqlfluff/sqlfluff#6119]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When enabling "parallel processing" mode with the `--processes N` switch, `JJ01` rule violations are not detected when passing a directory as the linting target. Other rule violations seem to be correctly detected.

### Expected Behaviour

Files where `JJ01` is violated are detected in parallel processing mode when directories are passed as the target.

### Observed Behaviour

`JJ01` rule violations are not detected when passing a directory as the linting target with the `--processes N` switch.

### Test setup

All these files are in a directory called `sqlfluff_bug_report`

## bad_lint_1.sql (Containing only a JJ01 violation)

```
SELECT 1
FROM {{source('test_source', 'test_model') }}
```

## bad_lint_2.sql (Containing other rule bundle violations)

```
SELECT 1, 2 as TWO
FROM {{source('test_source', 'test_model') }}
```

## good_lint.sql (Containing no violations)

```
SELECT 1
FROM {{ source('test_source', 'test_model') }}
```

## Full directory, single processing (correct behaviour)

```
╰─❯ sqlfluff lint sqlfluff_bug_report/ -v                                                                                                                                                                           ─╯
==== sqlfluff ====
sqlfluff:                3.1.1 python:                 3.8.18
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:               jinja
rules:                                all
==== readout ====

== [sqlfluff_bug_report/bad_lint_1.sql] FAIL
L:   2 | P:   6 | JJ01 | Jinja tags should have a single whitespace on either
                       | side: {{source('test_source', 'test_model') }}
                       | [jinja.padding]
== [sqlfluff_bug_report/bad_lint_2.sql] FAIL
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
L:   1 | P:   8 | AL03 | Column expression without alias. Use explicit `AS`
                       | clause. [aliasing.expression]
L:   1 | P:  13 | CP01 | Keywords must be upper case.
                       | [capitalisation.keywords]
L:   1 | P:  16 | CP02 | Unquoted identifiers must be lower case.
                       | [capitalisation.identifiers]
L:   2 | P:   6 | JJ01 | Jinja tags should have a single whitespace on either
                       | side: {{source('test_source', 'test_model') }}
                       | [jinja.padding]
== [sqlfluff_bug_report/good_lint.sql] PASS
==== summary ====
violations:        6 status:         FAIL
```

## Full directory, multi processing (wrong/buggy behaviour)

Note that `bad_lint_1.sql` is marked as a PASS and the JJ01 violation is missing from the linting of `bad_lint_2.sql`.

```
╰─❯ sqlfluff lint sqlfluff_bug_report/ -v --processes 4                                                                                                                                                             ─╯
==== sqlfluff ====
sqlfluff:                3.1.1 python:                 3.8.18
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:               jinja
rules:                                all
==== readout ====

effective configured processes:  4
== [sqlfluff_bug_report/bad_lint_1.sql] PASS
== [sqlfluff_bug_report/bad_lint_2.sql] FAIL
L:   1 | P:   1 | LT09 | Select targets should be on a new line unless there is
                       | only one select target. [layout.select_targets]
L:   1 | P:   8 | AL03 | Column expression without alias. Use explicit `AS`
                       | clause. [aliasing.expression]
L:   1 | P:  13 | CP01 | Keywords must be upper case.
                       | [capitalisation.keywords]
L:   1 | P:  16 | CP02 | Unquoted identifiers must be lower case.
                       | [capitalisation.identifiers]
== [sqlfluff_bug_report/good_lint.sql] PASS
==== summary ====
violations:        4 status:         FAIL

```

## Single file, single processing (correct behaviour)

```
╰─❯ sqlfluff lint sqlfluff_bug_report/bad_lint_1.sql -v                                                                                                                                                             ─╯
==== sqlfluff ====
sqlfluff:                3.1.1 python:                 3.8.18
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:               jinja
rules:                                all
==== readout ====

== [sqlfluff_bug_report/bad_lint_1.sql] FAIL
L:   2 | P:   6 | JJ01 | Jinja tags should have a single whitespace on either
                       | side: {{source('test_source', 'test_model') }}
                       | [jinja.padding]
==== summary ====
violations:        1 status:         FAIL
```

## Single file, multi processing (wrong/buggy behaviour)

```
╰─❯ sqlfluff lint sqlfluff_bug_report/bad_lint_1.sql -v --processes 4                                                                                                                                               ─╯
==== sqlfluff ====
sqlfluff:                3.1.1 python:                 3.8.18
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:               jinja
rules:                                all
==== readout ====

== [sqlfluff_bug_report/bad_lint_1.sql] FAIL
L:   2 | P:   6 | JJ01 | Jinja tags should have a single whitespace on either
                       | side: {{source('test_source', 'test_model') }}
                       | [jinja.padding]
==== summary ====
violations:        1 status:         FAIL
```

### Dialect

Snowflake

### Version

3.1.1

## .sqlfluff

```
[sqlfluff]
dialect = snowflake
exclude_rules = AL07,ST06
large_file_skip_byte_limit = 0
large_file_skip_char_limit = 0
max_line_length = 120
sql_file_exts = .sql
templater = jinja

[sqlfluff:indentation]
indent_unit = space
indented_joins = True
indented_using_on = True
tab_space_size = 2
template_blocks_indent = False

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:layout:type:comma]
spacing_before = touch
line_position = leading

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
library_path = sqlfluff_libs

[sqlfluff:templater:jinja:context]
target = { name = dev }
execute = false

[sqlfluff:rules]
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

[sqlfluff:rules:capitalisation.keywords]
# Keywords
capitalisation_policy = upper

[sqlfluff:rules:capitalisation.identifiers]
# Column identifiers must be lowercase for better readability
extended_capitalisation_policy = lower

[sqlfluff:rules:references.keywords]
# Keywords should not be used as identifiers
# https://docs.sqlfluff.com/en/stable/rules.html#rule-references.keywords
# But allow these exceptions
ignore_words = name,url,template,language,location,model,email,data,display_name,body,error,range,date,type,options,index,role,timestamp,account,action,network

[sqlfluff:rules:references.special_chars]
# Do not use special characters in identifiers.
# https://docs.sqlfluff.com/en/stable/rules.html#rule-references.special_chars
allow_space_in_identifier = False
# Allow Snowflake external table metadata pseudocolumns "metadata$filename" and "metadata$file_row_number"
ignore_words_regex = ^metadata\$file
# Some source columns may have spaces because they come from arbitrary CSVs, sheets etc.
quoted_identifiers_policy = column_aliases

[sqlfluff:rules:aliasing.forbid]
# Avoid table aliases in from clauses and join conditions
# https://docs.sqlfluff.com/en/stable/rules.html#rule-aliasing.forbid
force_enable = False
```

## .sqlfluffignore

```
macros/
target/
tests/
dbt_packages/
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
